argoverse/utils/sim2.py

```diff
@@ -0,0 +1,106 @@
+"""
+Utility for 2d rigid body transformations with scaling.
+
+Refs:
+    http://ethaneade.com/lie_groups.pdf
+    https://github.com/borglab/gtsam/blob/develop/gtsam_unstable/geometry/Similarity3.h
+"""
+
+from typing import Union
+
+import numpy as np
+
+from argoverse.utils.helpers import assert_np_array_shape
+
+
+class Sim2:
+    """ Implements the Similarity(2) class."""
+
+    def __init__(self, R: np.ndarray, t: np.ndarray, s: Union[int, float]) -> None:
+        """Initialize from rotation R, translation t, and scale s.
+
+        Args:
+            R: array of shape (2x2) representing 2d rotation matrix
+            t: array of shape (2,) representing 2d translation
+            s: scaling factor
+        """
+        assert_np_array_shape(R, (2, 2))
+        assert_np_array_shape(t, (2,))
+        assert isinstance(s, float) or isinstance(s, int)
+        self.R_ = R.astype(np.float32)
+        self.t_ = t.astype(np.float32)
+        self.s_ = float(s)
+
+    def __eq__(self, other: object) -> bool:
+        """Check for equality with other Sim(2) object"""
+        if not isinstance(other, Sim2):
+            return False
+
+        if not np.isclose(self.scale, other.scale):
+            return False
+
+        if not np.allclose(self.rotation, other.rotation):
+            return False
+
+        if not np.allclose(self.translation, other.translation):
+            return False
+
+        return True
+
+    @property
+    def rotation(self) -> np.ndarray:
+        """Return the 2x2 rotation matrix"""
+        return self.R_
+
+    @property
+    def translation(self) -> np.ndarray:
+        """Return the (2,) translation vector"""
+        return self.t_
+
+    @property
+    def scale(self) -> float:
+        """Return the scale."""
+        return self.s_
+
+    @property
+    def matrix(self) -> np.ndarray:
+        """Calculate 3*3 matrix group equivalent"""
+        T = np.zeros((3, 3))
+        T[:2, :2] = self.R_
+        T[:2, 2] = self.t_
+        T[2, 2] = 1 / self.s_
```

Comment:
We might want to validate that `self.s_` does not equal zero in the constructor to avoid `ZeroDivisionError`.

Reply:
very good point -- i just updated the constructor to throw an exception in that case. and i added a new unit test to check that case also